prover/src/logup_gkr/mod.rs

```diff
@@ -407,3 +422,66 @@ pub enum GkrProverError {
     #[error("failed to generate the random challenge")]
     FailedToGenerateChallenge,
 }
+
+// HELPER
+// =================================================================================================
+
+/// Computes the cumulative sum, also called prefix sum, of a vector of field elements.
+///
+/// Depending on whether `concurrent` feature is enabled, the function either uses a naive serial
+/// implementation or an implementation which makes use of parallelism.
+fn prefix_sum<E: FieldElement>(vector: &mut [E]) {
+    #[cfg(feature = "concurrent")]
+    prefix_sum_parallel(vector, 1024);
+
+    #[cfg(not(feature = "concurrent"))]
+    prefix_sum_truncate_left(vector, E::ZERO);
+}
+
+/// Computes the cumulative sum, also called prefix sum, of a vector of field elements using
+/// parallelism, in place.
+///
+/// The function divides the vector into non-overlapping segments and then computes an array of sums
+/// for each segment. The function then applies the naive serial implementation to each segment and
+/// uses the pre-computed sums in each segment in order to coordinate the results in the different
+/// segments.
+///
+/// The input vector is of the form `0 || values` where `values` are the values the cumulative sum
+/// vector will be computed for, in place.
+#[cfg(feature = "concurrent")]
+fn prefix_sum_parallel<E: FieldElement>(vector: &mut [E], batch_size: usize) {
+    use utils::{chunks, chunks_mut, iter, iter_mut};
+
+    let num_partitions = (vector.len() + batch_size - 1) / batch_size;
+    let mut sum_per_partition = vec![E::ZERO; num_partitions];
+
+    chunks!(vector, batch_size)
+        .zip(iter_mut!(sum_per_partition))
+        .for_each(|(chunk, entry)| *entry = chunk.iter().fold(E::ZERO, |acc, term| acc + *term));
```

Comment:
This will split things up into relatively small batches (length 1024 each), right? Not sure if it will be materially different, but I would probably split this into the number of partitions which is based on the number of threads.

Reply:
Good point! Updated to take the number of threads into account